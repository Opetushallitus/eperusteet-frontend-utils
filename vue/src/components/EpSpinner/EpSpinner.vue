<template>
<div class="spinner" :class="{'small': small}">
  <div class="oph-spinner">
    <div class="oph-bounce oph-bounce1" :style="style"></div>
    <div class="oph-bounce oph-bounce2" :style="style"></div>
    <div class="oph-bounce oph-bounce3" :style="style"></div>
  </div>
</div>
</template>

<script lang="ts">
import { Vue, Component, Prop } from 'vue-property-decorator';

@Component
export default class EpSpinner extends Vue {
  @Prop({ required: false, type: Boolean })
  private small!: boolean;

  @Prop({ required: false })
  private color!: string;

  get style() {
    if (this.color) {
      return { 'background-color': this.color + ' !important' };
    }
  }
}
</script>

<style scoped lang="scss">
@import "../../styles/_variables.scss";
@import "./style.scss";

.spinner {
  margin-top: 5px;

  &.small {
    .oph-bounce {
      height: 10px;
      width: 10px;
    }
  }

  .oph-bounce {
    background-color: $spinner-background-color !important;
  }
}
</style>
