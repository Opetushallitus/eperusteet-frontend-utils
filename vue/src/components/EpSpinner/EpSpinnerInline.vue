<template lang="pug">
.oph-spinner.oph-spinner-in-button
  .oph-bounce.oph-bounce1
  .oph-bounce.oph-bounce2
  .oph-bounce.oph-bounce3
</template>

<style scoped lang="scss">
.oph-spinner-in-button {
  margin-left: 15px;
}
</style>
