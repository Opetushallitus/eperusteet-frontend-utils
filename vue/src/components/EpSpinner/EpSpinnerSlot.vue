<template>
<div class="spinner">
  <ep-spinner v-if="isLoading"/>
  <slot v-else></slot>
</div>
</template>

<script lang="ts">
import { Vue, Prop, Component } from 'vue-property-decorator';
import EpSpinner from './EpSpinner.vue';

@Component({
  components: {
    EpSpinner,
  },
})
export default class EpSpinnerSlot extends Vue {
  @Prop({ required: true, type: Boolean })
  private isLoading!: boolean;
}
</script>

<style>
</style>
