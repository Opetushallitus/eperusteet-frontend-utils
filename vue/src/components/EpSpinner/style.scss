@keyframes bounceDelay {
  0%,
  80%,
  100% {
    transform: scale(0);
  }

  40% {
    transform: scale(1);
  }
}

.oph-spinner {
  width: 100px;
  margin-right: auto;
  margin-left: auto;
  text-align: center;
}

.oph-spinner-in-button {
  display: inline-block;
  width: auto;
  margin-right: 0.5rem;
}

.oph-spinner .oph-bounce {
  width: 18px;
  height: 18px;
  border-radius: 100%;
  background-color: --color-primary;
  display: inline-block;
  animation: bounceDelay 1.4s infinite ease-in-out;

  /* Prevent first frame from flickering when animation starts */
  animation-fill-mode: both;
}

.oph-spinner-in-button .oph-bounce {
  display: inline-block;
  width: 0.75rem;
  height: 0.75rem;
  background-color: #fff;
}

.oph-spinner .oph-bounce1 {
  animation-delay: -0.32s;
}

.oph-spinner .oph-bounce2 {
  animation-delay: -0.16s;
}
