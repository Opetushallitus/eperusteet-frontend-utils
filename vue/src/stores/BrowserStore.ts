import Vue from 'vue';
import VueCompositionApi, { reactive, computed } from '@vue/composition-api';
import { NavigationNodeDto, PerusteprojektiDto, PerusteDto, Ulkopuoliset, Perusteprojektit, Perusteet } from '@shared/api/eperusteet';
import _ from 'lodash';

Vue.use(VueCompositionApi);

const state = reactive({
  window: {
    width: window.innerWidth,
    height: window.innerHeight,
  },
  scrollY: 0,
  focused: null as FocusEvent | null,
  latestKeypress: null as any | null,
});

document.addEventListener('scroll', (ev) => {
  state.scrollY = window.pageYOffset;
});

window.addEventListener('resize', (ev) => {
  const { innerWidth, innerHeight } = ev.target as any;
  state.window.width = innerWidth;
  state.window.height = innerHeight;
});

document.addEventListener('focusin', (ev) => {
  if (ev.target) {
    state.focused = ev as any;
  }
});

document.addEventListener('focusout', (ev) => {
  state.focused = null;
});

document.addEventListener('keyup', (ev) => {
  if (!ev) {
    return;
  }

  const press = _.pick(ev, 'ctrlKey', 'code', 'shiftKey', 'type', 'keyCode');
  state.latestKeypress = press;
});

export class BrowserStore {
  public readonly scrollY = computed(() => state.scrollY);
  public readonly window = computed(() => state.window);
  public readonly focused = computed(() => state.focused);
  public readonly activeElement = computed(() => state.focused?.target);
  public readonly latestKeypress = computed(() => state.latestKeypress);
  public readonly navigationVisible = computed(() => state.window.width > 991);
}
