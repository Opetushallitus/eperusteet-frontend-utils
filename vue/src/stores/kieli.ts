import { createLogger } from '../utils/logger';
import { Kieli } from '../tyypit';
import _ from 'lodash';
import VueI18n from 'vue-i18n';
import Vue, { VueConstructor } from 'vue';
import moment from 'moment';
import 'moment/locale/en-gb';
import 'moment/locale/fi';
import 'moment/locale/ru';
import 'moment/locale/se';
import 'moment/locale/sv';
import VueCompositionApi, { computed, reactive } from '@vue/composition-api';
import { updateRelativeTime } from '../plugins/aikaleima';
import { Computed } from '../utils/interfaces';

Vue.use(VueCompositionApi);

declare module 'vue/types/vue' {
  interface Vue {
    $locale: Computed<string>;
    $slang: Computed<string>;
    $t: typeof VueI18n.prototype.t;
    $tc: typeof VueI18n.prototype.tc;
    $te: typeof VueI18n.prototype.te;
  }
}

export const UiKielet = Object.freeze(['fi', 'sv', 'en']);
const logger = createLogger('Kieli');

function getMessages() {
  const result: any = {};
  try {
    result.fi = require('@/translations/locale-fi.json');
  }
  catch (e) {}

  try {
    result.sv = require('@/translations/locale-sv.json');
  }
  catch (e) {}
  return result;
}

export class KieliStore {
  private vi18n!: VueI18n;

  public get i18n() {
    return this.vi18n;
  }

  /**
   * Add language support to root vue instance.
   *
   */
  public install(v: VueConstructor, config = {} as Partial<VueI18n.I18nOptions>) {
    moment.locale(Kieli.fi);
    this.vi18n = new VueI18n({
      fallbackLocale: Kieli.fi,
      locale: Kieli.fi,
      messages: getMessages(),
      ..._.cloneDeep(config),
    });
    v.prototype.$locale = this.uiKieli;
    v.prototype.$slang = this.sisaltoKieli;
    moment.updateLocale(Kieli.fi, updateRelativeTime());
  }

  /**
   * Load translation from other source
   *
   */
  public async load(kaannokset: Kaannokset) {
    logger.info('Initing locales');
    const results = await this.fetchLocaleMap(kaannokset);
    _.forEach(results, (locales, lang) => {
      this.i18n!.mergeLocaleMessage(lang, locales);
    });
  }

  private readonly state = reactive({
    sisaltoKieli: Kieli.fi,
  });

  public readonly getUiKieli = computed(() => this.i18n?.locale);
  public readonly getSisaltoKieli = computed(() => this.state.sisaltoKieli);
  public readonly sisaltoKieli = computed(() => this.state.sisaltoKieli);
  public readonly uiKieli = computed(() => this.i18n?.locale);

  public readonly getAikakaannokset = computed(() => {
    const kieli = this.state.sisaltoKieli;
    return {
      days: _.map(moment.weekdays(), (day: string) => _.toUpper(_.first(day))),
      months: moment.monthsShort(),
      placeholder: {
        date: this.i18n!.t('valitse-pvm'),
        dateRange: this.i18n!.t('valitse-pvm-jana'),
      },
    };
  })

  public setUiKieli(kieli: Kieli) {
    if (this.i18n!.locale !== kieli && _.includes(UiKielet, kieli)) {
      moment.locale(kieli);
      this.i18n!.locale = kieli;
    }
  }

  public setSisaltoKieli(kieli: Kieli) {
    if (this.state.sisaltoKieli !== kieli && _.includes(UiKielet, kieli)) {
      this.state.sisaltoKieli = kieli;
    }
  }

  public haeLokalisoituOlio(avain: string) {
    const result = {
      fi: this.i18n!.t(avain, 'fi'),
      sv: this.i18n!.t(avain, 'sv'),
      en: this.i18n!.t(avain, 'en'),
    };
    return result;
  }

  /**
   * Check if text query matches object
   *
   */
  public search(query: string, text: any, config = {}): boolean {
    if (text && query) {
      const target = _.isString(text) ? text : text[this.getSisaltoKieli.value];
      return _.includes(_.toLower(target), _.toLower(query));
    }
    else {
      return true;
    }
  }

  public filterBy(field: string, query: string, config = {}) {
    return (data: any) => this.search(query, data[field]);
  }

  public searchFn(query: string) {
    return (text: any) => this.search(query, text);
  }

  /**
   * i18n locale convert
   *
   */
  public t(value: string): string {
    if (this.vi18n) {
      return this.vi18n.t(value) as string || '<' + value + '>';
    }
    else {
      return value;
    }
  };

  /**
   * Convert field into sortable value
   *
   * @returns {string}
   */
  public sortValue(value: any): string {
    if (!value) {
      return '';
    }
    else if (_.isObject(value)) {
      const locale = this.getSisaltoKieli.value;
      return (value as any)[locale];
    }
    else {
      return value;
    }
  };

  public kaanna(value: LokalisoituTeksti | undefined | null): string {
    if (!value) {
      return '';
    }
    else if (_.isObject(value)) {
      const locale = this.getSisaltoKieli.value;
      const kielet = [locale, ..._.pull(['fi', 'sv', 'en', 'se', 'ru'], locale)];
      let teksti = '';

      _.forEach(kielet, kieli => {
        if (!_.isEmpty(value[kieli])) {
          teksti = value[kieli] as string;
          return false;
        }
      });

      return teksti;
    }
    else {
      logger.warn('"$kaanna" on tekstiolioiden kääntämiseen. Käytä vue-i18n vastaavaa funktiota. Esimerkiksi "$t()".', 'Käännös:', '"' + value + '"');
      return value;
    }
  };

  public kaannaOlioTaiTeksti(value: LokalisoituTeksti | string): string {
    if (_.isObject(value)) {
      return this.kaanna(value);
    }
    else if (_.isString(value)) {
      return this.t(value);
    }
    else {
      return value;
    }
  }

  private async fetchLocaleMap(kaannokset: Kaannokset) {
    try {
      const result: any = {};
      const localeObj = kaannokset;
      _.forEach(localeObj, (locales, lang) => {
        result[lang] = {};
        if (locales) {
          for (const locale of locales) {
            if (locale.key && locale.value) {
              result[lang][locale.key] = locale.value;
            }
          }
        }
      });
      return result;
    }
    catch (err) {
      logger.error('Käännösten haku epäonnistui', err.message);
      return {};
    }
  }
}

export const Kielet = new KieliStore();

export interface Kaannos {
  key: string;
  value: string;
}

export interface Kaannokset {
  fi?: Kaannos[];
  sv?: Kaannos[];
  en?: Kaannos[];
}

export interface LokalisoituTeksti {
  [locale: string]: string | null | undefined | number;
  _id?: number;
  _tunniste?: string;
}
