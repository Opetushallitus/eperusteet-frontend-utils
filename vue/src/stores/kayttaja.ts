import _ from 'lodash';
import Vue from 'vue';
import { Store, Getter, State } from '@shared/stores/store';
import { Kayttajat as KayttajatApi, KayttajanTietoDto } from '@shared/api/eperusteet';
import { createLogger } from '@shared/utils/logger';
import VueCompositionApi, { reactive, computed, ref, watch } from '@vue/composition-api';

Vue.use(VueCompositionApi);

// FIXME: tyypitä backendiin
export type Oikeus = 'luku' | 'kommentointi' | 'muokkaus' | 'luonti' | 'poisto' | 'tilanvaihto' | 'hallinta';
export type OikeusKohde = 'opetussuunnitelma' | 'pohja';
export interface Oikeudet { [kohde: string]: Oikeus[]; }

function getOikeusArvo(oikeus: Oikeus) {
  switch (oikeus) {
  case 'luku': return 1;
  case 'kommentointi': return 2;
  case 'muokkaus': return 3;
  case 'luonti': return 4;
  case 'poisto': return 5;
  case 'tilanvaihto': return 6;
  default: return 0;
  }
}

export function parsiEsitysnimi(tiedot: any): string {
  if (tiedot.kutsumanimi && tiedot.sukunimi) {
    return tiedot.kutsumanimi + ' ' + tiedot.sukunimi;
  }
  else {
    return tiedot.oidHenkilo as string;
  }
}

const logger = createLogger('Kayttaja');

export class KayttajaStore {
  public state = reactive({
    organisaatiot: [] as any[],
    tiedot: {} as KayttajanTietoDto,
    virkailijat: [] as any[],
    oikeudet: {
      opetussuunnitelma: [],
      pohja: [],
    } as Oikeudet,
  });

  public readonly organisaatiot = computed(() => this.state.organisaatiot);
  public readonly tiedot = computed(() => this.state.tiedot);
  public readonly virkailijat = computed(() => this.state.virkailijat);
  public readonly oikeudet = computed(() => this.state.oikeudet);
  public readonly nimi = computed(() => parsiEsitysnimi(this.state.tiedot));

  public async init() {
    try {
      logger.info('Haetaan käyttäjän tiedot');
      this.state.tiedot = (await KayttajatApi.getKirjautunutKayttajat()).data;
      logger.info('Käyttäjän tiedot', this.tiedot.value);
    }
    catch (err) {
      logger.error('Käyttäjän tietojen lataus epäonnistui', err.message);
    }
  }

  public async hasOikeus(oikeus: Oikeus, kohde: OikeusKohde = 'opetussuunnitelma') {
    if (!oikeus) {
      return false;
    }
    else if (oikeus === 'hallinta') {
      return this.hasHallintaoikeus(kohde);
    }
    else {
      return this.vertaa(oikeus, kohde);
    }
  }

  private vertaa(oikeus: Oikeus, kohde: OikeusKohde = 'opetussuunnitelma') {
    const haettu = getOikeusArvo(oikeus);
    if (haettu === 0) {
      return false;
    }
    else {
      const korkein = _.max(_.map(this.oikeudet.value[kohde], getOikeusArvo)) || 0;
      return korkein >= haettu;
    }
  }

  private hasHallintaoikeus(kohde) {
    return _.includes(this.oikeudet.value[kohde], 'luonti');
  }
}

export const Kayttajat = new KayttajaStore();
