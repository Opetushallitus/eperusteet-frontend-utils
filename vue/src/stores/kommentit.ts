import _ from 'lodash';
import { Kommentointi } from '@shared/api/ylops';
import Vue from 'vue';

import VueCompositionApi, { reactive, computed, ref, watch } from '@vue/composition-api';
import VueScrollTo from 'vue-scrollto';
import { Kielet } from '@shared/stores/kieli';

import { unwrap } from '@shared/utils/wraps';

import { createLogger } from '@shared/utils/logger';
Vue.use(VueCompositionApi);
const logger = createLogger('Kayttaja');

export const UusiKommenttiHandle = 'uusi-kommentti';

export function parsiEsitysnimi(tiedot: any): string {
  if (tiedot.kutsumanimi && tiedot.sukunimi) {
    return tiedot.kutsumanimi + ' ' + tiedot.sukunimi;
  }
  else {
    return tiedot.oidHenkilo as string;
  }
}

export function nestedRemove<T>(
  obj: T,
  condition: ((val: T) => boolean) | object,
  nestedField: string): T {
  return {
    ...obj,
    [nestedField]: _.chain(obj[nestedField])
      .reject(condition)
      .map(v => nestedRemove(v, condition, nestedField))
      .value(),
  };
}

export function nestedMap<T>(
  obj: T,
  nestedField: string,
  mapFn: ((val: T) => T)): T {
  return {
    ...obj,
    [nestedField]: _.map(obj[nestedField], mapFn),
  };
}

const CommentStyles = document.createElement('style');
document.head.appendChild(CommentStyles);
let ActiveCommentStyleIdx = -1;

class KommenttiStore {
  private state = reactive({
    threadUuid: null as string | null,
    thread: null as any | null,
    isActive: false,
    isLoading: false,
    selection: false,
    bounds: null as any | null,
    visibleChains: [] as string[], // List of visible UUIDs
    activeThreads: [] as any[], // Comments of visible chains
  });

  public readonly threadUuid = computed(() => this.state.threadUuid);
  public readonly thread = computed(() => _.sortBy(this.state.thread, 'luotu'));
  public readonly isLoading = computed(() => this.state.isLoading);
  public readonly hasSelection = computed(() => this.state.selection);
  public readonly bounds = computed(() => this.state.bounds);
  public readonly visibleThreads = computed(() => this.state.visibleChains || []);
  public readonly activeThreads = computed(() => this.state.activeThreads || []);

  public readonly surrounding = computed(() => {
    if (this.state.threadUuid) {
      const idx = _.findIndex(this.state.visibleChains, c => c === this.state.threadUuid);
      if (idx >= 0) {
        const previousIdx = idx === 0 ? _.size(this.state.visibleChains) - 1 : idx - 1;
        const nextIdx = (idx + 1) % _.size(this.state.visibleChains);
        return {
          previous: this.state.visibleChains[previousIdx],
          next: this.state.visibleChains[nextIdx],
        };
      }
    }
    return {
      previous: null,
      next: null,
    };
  });

  constructor() {
    const onChange = (ev) => {
      const selection = document.getSelection();
      this.state.selection = !selection || !selection?.isCollapsed;
      if (selection && this.state.selection) {
        const range = selection.getRangeAt(selection.rangeCount - 1);
        this.state.bounds = range.getBoundingClientRect();
      }
      else {
        this.state.bounds = null;
      }
    };

    document.onselectionchange = (ev) => {
      onChange(ev);
    };
  }

  public setActive(value: boolean) {
    this.state.isActive = value;
    if (value) {
      this.state.activeThreads = [];
      this.updateVisibleThreads();
    }
    else {
      this.state.activeThreads = [];
    }
  }

  private clearCommentStyle() {
    if (ActiveCommentStyleIdx >= 0) {
      const sheet = CommentStyles.sheet as CSSStyleSheet;
      if (sheet) {
        try {
          sheet.deleteRule(ActiveCommentStyleIdx);
        }
        catch (err) { }
      }
    }
  }

  private watchVisibleThreads = watch(this.visibleThreads, async () => {
    if (!_.isEmpty(this.visibleThreads.value)) {
      const res = await Promise.all(_(this.visibleThreads.value)
        .reject(uuid => uuid === UusiKommenttiHandle)
        .map(uuid => Kommentointi.getKommenttiByKetjuUuid(uuid))
        .value());
      this.state.activeThreads = _.map(res, 'data');
    }
  });

  private onLangChange = watch([Kielet.sisaltoKieli], () => {
    this.state.threadUuid = null;
    this.state.thread = null;
    this.clearCommentStyle();
  });

  private watcher = watch(this.threadUuid, async () => {
    if (this.threadUuid.value && this.threadUuid.value !== UusiKommenttiHandle) {
      const thread = await Kommentointi.getKommenttiByKetjuUuid(this.threadUuid.value);
      this.state.thread = thread.data;
    }
  });

  public async clearThread() {
    this.clearCommentStyle();
    this.state.threadUuid = null;
    this.state.thread = null;
    VueScrollTo.scrollTo('#keskustelu-sisalto', 300);
  }

  private obs: MutationObserver | null = null;

  public async activateThread(uuid: string) {
    this.state.thread = null;

    if (uuid === UusiKommenttiHandle) {
      return;
    }

    if (this.state.isLoading) {
      logger.info('Still loading', uuid);
      return;
    }

    if (uuid === UusiKommenttiHandle) {
      this.state.threadUuid = UusiKommenttiHandle;
      return;
    }

    logger.info('activating thread', uuid);

    this.clearCommentStyle();
    ActiveCommentStyleIdx = (CommentStyles.sheet as CSSStyleSheet).insertRule(`span[kommentti="${uuid}"] {
      background-color: #ffd900 !important;
    }`);

    this.state.isLoading = true;
    this.state.threadUuid = uuid;

    try {
      const thread = await Kommentointi.getKommenttiByKetjuUuid(uuid);
      this.state.thread = thread.data;
      logger.info('thread found', this.thread.value);
      return true;
    }
    catch (err) {
      logger.info('could not activate thread', uuid);
      this.state.thread = {
        tunniste: uuid,
        kommentit: [],
      };
    }
    finally {
      this.state.isLoading = false;
    }
  }

  public async lisaaKahva(kommentti: any) {
    const ketju = await Kommentointi.addKetju(kommentti);
    this.state.threadUuid = ketju.data.thread as string;
    return ketju.data;
  }

  public async tallenna(kommentti) {
    if (kommentti.tunniste) {
      logger.info('updating comment', kommentti.tunniste);
      const result = (await Kommentointi.updateKommentti2019(kommentti.thread, kommentti)).data;
      this.state.thread = [result, ..._.reject(this.state.thread, c => c.tunniste === kommentti.tunniste)];
      return result;
    }
    else {
      logger.info('adding comment to', kommentti.parent);
      const result = (await Kommentointi.addKommentti2019(kommentti.thread, kommentti)).data;
      this.state.thread = [result, ..._.filter(this.state.thread, c => c.tunniste)];
      return result;
    }
  }

  public async poista(tunniste: string) {
    logger.info('removing comment', tunniste);
    await Kommentointi.poistaKommenttiKetju2019(tunniste);
    this.state.thread = _.reject(this.state.thread, c => c.tunniste === tunniste);
    unwrap(document.querySelector(`span[kommentti="${tunniste}"]`));
    this.clearCommentStyle();
  }

  detach() {
    if (this.obs) {
      this.obs.disconnect();
      this.state.thread = null;
      this.obs = null;
    }
  }

  public readonly updateVisibleThreads = _.debounce(() => {
    const chains = [] as string[];
    document.querySelectorAll('[kommentti]').forEach(k => {
      const uuid = k.getAttribute('kommentti');
      if (uuid) {
        chains.push(uuid);
      }
    });
    logger.info('Updating visible threads', chains);
    this.state.visibleChains = chains;
  }, 300);

  public findTekstikappaleNode(node: Node) {
    function isEditorContent(n: Node) {
      if ((el as any)?.__vue__?.$options?._componentTag === 'editor-content') {
        const value = (el as any)?.__vue__?.$parent?.value;
        const tekstiId = Number(value?._id);
        if (tekstiId) {
          return true;
        }
      }
      else {
        return false;
      }
    }

    let el = node.parentNode;
    while (el !== null && el !== el.parentNode && !isEditorContent(el)) {
      el = el.parentNode;
    }

    if (el) {
      return {
        el,
        tekstiId: Number((el as any)?.__vue__?.$parent?.value?._id),
      };
    }
    else {
      return null;
    }
  }

  scrollTo(uuid: string) {
    VueScrollTo.scrollTo(`span[kommentti="${uuid}"]`, 300);
  }

  attach(el: Element) {
    const store = this;
    if (this.obs) {
      this.detach();
    }

    const mountCommentThreads = async (targets: Element[]) => {
      for (const thread of targets) {
        if (thread) {
          thread.addEventListener('click', () => {
            const uuid = thread.getAttribute('kommentti');
            if (uuid) {
              this.activateThread(uuid);
            }
          });
        }
      }
    };

    this.obs = new MutationObserver((mutations) => {
      for (const mutation of mutations) {
        mountCommentThreads((mutation.target as any).querySelectorAll('[kommentti]'));
      }
      this.updateVisibleThreads();
    });

    mountCommentThreads((document as any).querySelectorAll('[kommentti]'));

    this.obs.observe(el, {
      attributeFilter: ['kommentti'],
      attributeOldValue: false,
      attributes: true,
      characterData: false,
      characterDataOldValue: false,
      childList: true,
      subtree: true,
    });
  }
}

export const Kommentit = new KommenttiStore();
