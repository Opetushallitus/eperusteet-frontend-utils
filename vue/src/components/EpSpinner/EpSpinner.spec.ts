import { mount, createLocalVue, shallowMount } from '@vue/test-utils';
import EpSpinner from './EpSpinner.vue';
import VueI18n from 'vue-i18n';
import { Kielet } from '../../stores/kieli';
import Vue from 'vue';
import BootstrapVue from 'bootstrap-vue';

Vue.use(BootstrapVue);

describe('EpSpinner component', () => {
  const localVue = createLocalVue();

  test('Renders', async () => {
    const wrapper = mount(EpSpinner);
    expect(wrapper.html()).toMatchSnapshot();
  });
});
