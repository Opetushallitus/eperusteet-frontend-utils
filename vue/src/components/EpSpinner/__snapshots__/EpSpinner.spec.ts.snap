// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EpSpinner component Renders 1`] = `
<div class="spinner">
  <div class="oph-spinner">
    <div class="oph-bounce oph-bounce1"></div>
    <div class="oph-bounce oph-bounce2"></div>
    <div class="oph-bounce oph-bounce3"></div>
  </div>
</div>
`;
